many: end-to-end support for the bare base snap

mvo5 opened the pull request:
This branch adds the missing pieces to enable the "bare" base snap with a proper integration test.

The tests will require that the following snapcraft features get merged:
https://github.com/snapcore/snapcraft/pull/1420
https://github.com/snapcore/snapcraft/pull/1419

zyga commented on cmd/snap-confine/mount-support.c:
```diff
@@ -333,11 +334,34 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 						 scratch_dir, dir);
-				sc_do_mount(src, dst, NULL, MS_BIND, NULL);
-				sc_do_mount("none", dst, NULL, MS_SLAVE, NULL);
+				if (access(src, F_OK) == 0
```
We should be extra careful with symlinks here (the base snap may contain symlinks that will "pass" the F_OK test. I'd suggest using `lstat` instead, which should not be much longer but will give us the predictability.

mvo5 replied:
I'm find with lstat(), I kept it in sync with the lines before that use access() as well.

mvo5 replied:
Changed now

zyga commented on cmd/snap-confine/mount-support.c:
```diff
@@ -333,11 +334,34 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 	}
+	if (config->uses_base_snap) {
+		// snapd dir
+		const char *src = "/usr/lib/snapd";
```
This will clash with libexecdir layout.

mvo5 replied:
Indeed, I changed the code now to use the libexecdir configure option. However I think this needs to be even smarter, i.e. we need to take re-exec into account as well. I added a fixme. we could of course mount `dirname(sys.argv[0])` which is probably the easiest and most sensible thing to do.

zyga commented on cmd/snap-confine/mount-support.c:
```diff
@@ -217,6 +217,7 @@ struct sc_mount_config {
 	// The struct is terminated with an entry with NULL path.
 	const struct sc_mount *mounts;
 	bool on_classic_distro;
+	bool uses_base_snap;
```
I'd like to have code that makes this work without any flags. In essence all snaps use base snap so our approach should be more universal.

mvo5 replied:
The flag is currently only used to ensure `/usr/lib/snapd` (and later `snapctl`) is available for the snap. We need /usr/lib/snapd/snap-exec to actually running anything in a base so that is important. On a normal core system we do not have to make /usr/lib/snapd available, its already there.

zyga commented on cmd/snap-confine/mount-support.c:
```diff
@@ -333,11 +334,34 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
+		sc_do_mount(src, dst, NULL, MS_REC | MS_BIND, NULL);
+		sc_do_mount("none", dst, NULL, MS_REC | MS_SLAVE, NULL);
+
+		// FIXME: snapctl tool - our apparmor policy wants it in
```
How about this: let's always use `/var/lib/snapd/tools` as a place where snapd keeps it tooling. Snap confine could then mount tmpfs there and create a symlink or bind mount farm from each tool (wherever it is, even from the file that is invisible in the base snap) to an empty file here.

As long as we can require /var/lib/snapd to exist in base snaps we could then run our tools reliably.

sergiusens commented on tests/lib/snaps/test-snapd-base-bare/meta/snap.yaml:
```diff
@@ -0,0 +1,4 @@
+name: test-snapd-base-bare
+version: 1.0
+type: base
```
this new type isn't added in any snapcraft PR, only the `base` field, but not the `base` snap `type`

mvo5 commented:
@zyga I addressed your comments and replied. I would love to have this as an early preview for bases in 2.28. It would allow e.g. Neal to start experimenting with the fedora core snap for real.

mvo5 commented on cmd/snap-confine/mount-support.c:
```diff
@@ -43,14 +43,16 @@
 
 #define MAX_BUF 1000
 
+#ifndef LIBEXECDIR
```
Just paranoia, I can remove it if it bothers you.

mvo5 commented:
@zyga MS_REC is now removed

niemeyer commented on cmd/snap-confine/mount-support.c:
```diff
@@ -43,14 +43,16 @@
 #define MAX_BUF 1000
 
+#ifndef LIBEXECDIR
+LIBEXECDIR = "/usr/lib/snapd"
```
Pretty sure this won't compile (I hope it won't!).

mvo5 replied:
Haha - you are very right, silly me!

niemeyer commented on cmd/snap-confine/mount-support.c:
```diff
@@ -330,16 +333,49 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
+	if (config->uses_base_snap) {
+		// snapd dir: we need it so that we can actually run
+		// anything because snap-exec needs to be visible the
+		// base snap
```
Seems slightly misleading. We don't want snap-exec to ship in base snaps, right?

mvo5 replied:
Indeed, I reworded the comment now to make more clear what is needed.

niemeyer commented on packaging/ubuntu-14.04/rules:
```diff
@@ -128,6 +128,13 @@ override_dh_auto_build:
 	dh_auto_build -- $(BUILDFLAGS) $(TAGS) $(GCCGOFLAGS)
+	# Generate static snap-exec - it somehow includes CGO so we must
+	# force a static build here. We need a static snap-exec inside
+	# the core snap because not all bases will have a libc
```
:heart:

niemeyer replied:
The base field is on the app snap itself, while the base type is on the underlying snap. These sound special enough to be worth having a new type for them. We want to hide them in the default listing, for example.

niemeyer commented on tests/lib/snaps/test-snapd-base-bare/meta/snap.yaml:
```diff
@@ -0,0 +1,4 @@
+name: test-snapd-base-bare
```
Can't comment on the empty files/dirs above, so adding here:

Wouldn't it be better to create them dynamically inside the test instead of having these in the tree?

mvo5 replied:
Yes, thank you! I added a snapcraft.yaml that creates them now instead of manually doing it here.
